### PR TITLE
[Core][Enable gcs scheduler 6/n] Turn on gcs scheduling feature flag

### PR DESCRIPTION
**Pull request by Chong-Li**
## Why are these changes needed?

This is the third split PR of #25075. We are trying to turn on the gcs scheduler feature flag, and make sure it passes all test cases. Main changes include:

1. When gcs detects long task pending, it triggers global gc if needed. This is similar to what raylets has been doing for deadlock resource detection.
2. For gcs actor scheduler, if the actor creation task is canceled with `SCHEDULING_CANCELLED_UNSCHEDULABLE` failure type, then it means the lease request has never been sent out yet. So we have to handle this case specially.
3. `AffinityWithBundleSchedulingPolicy` now supports avoiding gpu nodes.
4. Some PG wildcard functions are fixed.
5. Two test cases are skipped.

Currently, I'm **turning on** the gcs scheduler feature flag. When we finalize the review and make sure all tests pass, we can determine whether leave it turned on by default.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/common/ray_config_def.h
Comment (scv119) [inline src/ray/common/ray_config_def.h:568]: we need a roll out/benchmark plan before turn it on by default.

File: src/ray/gcs/gcs_server/gcs_actor_manager.cc
Comment (scv119) [inline src/ray/gcs/gcs_server/gcs_actor_manager.cc:811]: what's the side effect of calling `CancelActorInScheduling` if actor_scheduling_canceled is true?

File: src/ray/gcs/gcs_server/gcs_placement_group_scheduler.cc
Comment (scv119) [inline src/ray/gcs/gcs_server/gcs_placement_group_scheduler.cc:623]: looks like a bug? can we add a unit test for it?

File: src/ray/raylet/scheduling/policy/affinity_with_bundle_scheduling_policy.cc
Comment (scv119) [inline src/ray/raylet/scheduling/policy/affinity_with_bundle_scheduling_policy.cc:35]: can we create a helper function with unit test for it?

**Review by scv119 — CHANGES_REQUESTED** [review]

File: src/ray/gcs/gcs_server/gcs_actor_manager.cc
Comment (Chong-Li) [reply src/ray/gcs/gcs_server/gcs_actor_manager.cc:811]: In gcs actor scheduler, `actor_scheduling_canceled` is set true only when gcs' `ClusterTaskManager` ReplyCancelled with `SCHEDULING_CANCELLED_UNSCHEDULABLE`.  If we call `CancelActorInScheduling` in this case, we will remove the task from `ClusterTaskManager`'s pending queue (via calling`GcsActorManager::RemovePendingActor()`). Then when we call `work_it = work_queue.erase(work_it);` (at line 122 in `ClusterTaskManager::ScheduleAndDispatchTasks()`), we will run into a `double free` fatal failure. 

To solve this problem, we can just call `work_it = work_queue.erase(work_it);` before calling `ReplyCancelled(*, rpc::RequestWorkerLeaseReply::SCHEDULING_CANCELLED_UNSCHEDULABLE, *)`. But then we call `GcsActorManager::RemovePendingActor()`, we will not find the task from the pending queue. Then we will go to `GcsActorScheduler::CancelOnLeasing()` and get a RAY_CHECK failure (line 229 in `GcsActorScheduler::CancelOnLeasing()`), because this worker lease has never been sent out.

Anyway, this logic is complicated, and I feel the current change is the most straightforward and easy-to-understand way.

File: src/ray/gcs/gcs_server/gcs_placement_group_scheduler.cc
Comment (Chong-Li) [reply src/ray/gcs/gcs_server/gcs_placement_group_scheduler.cc:625]: `test_placement_group_reschedule_when_node_dead ` can actually cover this function, but I also just added a unit test for it.

File: src/ray/raylet/scheduling/policy/affinity_with_bundle_scheduling_policy.cc
Comment (Chong-Li) [reply src/ray/raylet/scheduling/policy/affinity_with_bundle_scheduling_policy.cc]: I just fixed and simplified this part. I'm not sure if it is worth a dedicated unit test because now it's quite simple. 

BTW, it seems there is not a unit test for `AffinityWithBundleSchedulingPolicy` yet. Let me know if I missed it. @larrylian

File: src/ray/common/ray_syncer/ray_syncer.cc
Comment (wumuzi520) [inline src/ray/common/ray_syncer/ray_syncer.cc:359]: The move operations will not affect variables with `const &`

File: src/ray/raylet/scheduling/policy/scheduling_options.h
Comment (wumuzi520) [inline src/ray/raylet/scheduling/policy/scheduling_options.h:89]: It's better to make `avoid_gpu_nodes` as a argument of `AffinityWithBundle` function.

File: src/ray/raylet/scheduling/policy/affinity_with_bundle_scheduling_policy.cc
Comment (wumuzi520) [inline src/ray/raylet/scheduling/policy/affinity_with_bundle_scheduling_policy.cc:35]: It's better to make this logic to a global util function.

File: src/ray/raylet/node_manager.cc
Comment (wumuzi520) [inline src/ray/raylet/node_manager.cc:1168]: Why delete this logic?

File: src/ray/gcs/gcs_server/gcs_server.cc
Comment (wumuzi520) [inline src/ray/gcs/gcs_server/gcs_server.cc:199]: Move the following logic to a method named `TryGlobgalGc()`?

File: src/ray/raylet/node_manager.cc
Comment (wumuzi520) [inline src/ray/raylet/node_manager.cc:1160]: Why move this logic at he very beginning of the method?

File: src/ray/gcs/gcs_server/gcs_server.cc
Comment (wumuzi520) [inline src/ray/gcs/gcs_server/gcs_server.cc:206]: It should be `++task_pending_schedule_detected_ > 0`?

File: src/ray/gcs/gcs_server/gcs_server.cc
Comment (wumuzi520) [inline src/ray/gcs/gcs_server/gcs_server.cc:193]: Is it right to * 1e9?

File: src/ray/raylet/scheduling/policy/scheduling_options.h
Comment (Chong-Li) [reply src/ray/raylet/scheduling/policy/scheduling_options.h:89]: Now `avoid_gpu_nodes` is a system-wise config, so we do the same thing with `Hybrid()` and `Spread()`. Making it as an argument of `AffinityWithBundle()` assumes it may change across different scheduling requests, which is not actually.

File: src/ray/common/ray_syncer/ray_syncer.cc
Comment (Chong-Li) [reply src/ray/common/ray_syncer/ray_syncer.cc:359]: Will do.

File: src/ray/raylet/node_manager.cc
Comment (Chong-Li) [reply src/ray/raylet/node_manager.cc:1160]: The head node is not in the local `ClusterResourceManager`'s resource view. So if there is a global gc signal sent from  the head node, it will be blocked by the `if (!cluster_resource_scheduler_->GetClusterResourceManager().UpdateNode())` statement.

File: src/ray/gcs/gcs_server/gcs_server.cc
Comment (Chong-Li) [reply src/ray/gcs/gcs_server/gcs_server.cc:206]: `++task_pending_schedule_detected_ > 0` would trigger the global gc the first time finding pending tasks in gcs. That might be overly reactive.

File: src/ray/gcs/gcs_server/gcs_server.cc
Comment (Chong-Li) [reply src/ray/gcs/gcs_server/gcs_server.cc:193]: It should work because `NodeManager` does it the same way.

File: src/ray/gcs/gcs_server/gcs_server.cc
Comment (wumuzi520) [inline src/ray/gcs/gcs_server/gcs_server.cc:766]: Why not use `ray_syncer_->BroadcastMessage(msg);` directly?

Comment (Chong-Li) [reply src/ray/gcs/gcs_server/gcs_server.cc:766]: Now `BroadcastMessage()` is a private function.

File: src/ray/common/ray_syncer/ray_syncer.cc
Comment (wumuzi520) [inline src/ray/common/ray_syncer/ray_syncer.cc:358]: BroadcastRaySyncMessage

File: src/ray/gcs/gcs_server/gcs_server.h
Comment (wumuzi520) [inline src/ray/gcs/gcs_server/gcs_server.h:273]: `int task_pending_schedule_detected_ = 0`

File: src/ray/gcs/gcs_server/gcs_actor_manager.cc
Comment (Chong-Li) [reply src/ray/gcs/gcs_server/gcs_actor_manager.cc]: The part has been moved to `ClusterTaskManager::ScheduleAndDispatchTasks`, which makes it much simpler. @scv119

File: src/ray/gcs/gcs_server/gcs_server.cc
Comment (wumuzi520) [inline src/ray/gcs/gcs_server/gcs_server.cc:760]:
```
auto msg = std::make_shared<syncer::RaySyncMessage>
...
ray_syncer_->BroadcastRaySyncMessage(msg);
```

File: src/ray/common/ray_syncer/ray_syncer.h
Comment (wumuzi520) [inline src/ray/common/ray_syncer/ray_syncer.h:144]: void BroadcastRaySyncMessage(std::shared_ptr<const RaySyncMessage> message);

File: src/ray/common/ray_config_def.h
Comment (Chong-Li) [reply src/ray/common/ray_config_def.h:593]: Yes. It is only used to test all cases with gcs scheduler enabled. I'll turn if off after code review.

**Review by wumuzi520 — APPROVED** [review]
LGTM!
@scv119 @JiahaoYao Could you review again?

File: src/ray/common/ray_config_def.h
Comment (scv119) [inline src/ray/common/ray_config_def.h:593]: let's not enable by default?

**Review by scv119 — APPROVED** [review]
Thanks! let's not enable it by default but let user opt in first?

**Comment by rkooo567** [comment]
should I merge it? There seems to be some test failures. Also, after we merge, can we try running large scale test with this flag on? cc @iycheng

**Review by rkooo567 — COMMENTED** [review]
Btw, in the current actor scheduling, we sort of try colocating actors to the node that submits it (e.g., driver). Is this behavior preserved with the new scheduler? I remember this caused regression without this before.

Also I will have one last review tomorrow. If I don't do it by tmrw, I will just merge it!

**Comment by Chong-Li** [comment]
> Btw, in the current actor scheduling, we sort of try colocating actors to the node that submits it (e.g., driver). Is this behavior preserved with the new scheduler? I remember this caused regression without this before.

Now gcs-based actor scheduler does not prefer the owner's node. If we feel this feature is needed in any scheduler, I could do it in another PR. There are still plenty of optimizations can be done.

**Comment by scv119** [comment]
ping @rkooo567

**Comment by rkooo567** [comment]
> Now gcs-based actor scheduler does not prefer the owner's node. If we feel this feature is needed in any scheduler, I could do it in another PR. There are still plenty of optimizations can be done.

We should do this before we turn this on by default! I've seen rllib related regression due to change of this behavior before. I don't have time to review this PR, so I will rely on other approvals..

**Pull request merged**
